klog.go

```diff
@@ -748,6 +748,49 @@ func (l *loggingT) printWithFileLine(s severity, logr logr.InfoLogger, file stri
 	l.output(s, logr, buf, file, line, alsoToStderr)
 }
 
+// printS if loggr is specified, no need to output with logging module. If
+// err arguments is specified, will call logr.Error, or output to errorLog severity
+func (l *loggingT) printS(err error, loggr logr.Logger, msg string, keysAndValues ...interface{}) {
+	if loggr != nil {
+		if err != nil {
+			loggr.Error(err, msg, keysAndValues)
+		} else {
+			loggr.Info(msg, keysAndValues)
+		}
+		return
+	}
+	b := &bytes.Buffer{}
+	b.WriteString(fmt.Sprintf("%q", msg))
+	if err != nil {
+		b.WriteByte(' ')
+		b.WriteString(fmt.Sprintf("err=%q", err.Error()))
+	}
+	kvListFormat(b, keysAndValues...)
+	var s severity
+	if err == nil {
+		s = infoLog
+	} else {
+		s = errorLog
+	}
+	l.printDepth(s, logging.logr, 1, b)
+}
+
+const missingValue = "(MISSING)"
+
+func kvListFormat(b *bytes.Buffer, keysAndValues ...interface{}) {
+	for i := 0; i < len(keysAndValues); i += 2 {
+		var v interface{}
+		k := keysAndValues[i]
+		if i+1 < len(keysAndValues) {
+			v = keysAndValues[i+1]
+		} else {
+			v = missingValue
+		}
+		b.WriteByte(' ')
+		b.WriteString(fmt.Sprintf("%v=%#v", k, v))
```

Comment:
This doesn't handle non-string keys.

Reply:
I don't think the KEP specified what we would print in case of a non-primitive values.  I guess %#v satisfies that, but something like JSON formatting seems better in every way, no?

https://play.golang.org/p/lkyMaj6OvYg

Reply:
Ok, but json format  may be support with next few issues,  if we use JSON as value format here, is there little confusion？

Reply:
@thokin, this format is intended as replacement of format strings ("%s", "%v", "%d") used in klog.Infof. InfoS format should not be to much worse that manually picked format in Infof. 

I'm for using "%#v" in this PR as we are still in Alpha stage of effort. We can think about proposing improvements when we start collect feedback in Beta.

Reply:
@thockin